docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+# Source Generators
+
+
+## Summary
+
+Source generators aim to enable _compile time metaprogramming_, that is, code that can be created
+at compile time and added to the compilation. Source generators will be able to read the contents of 
+the compilation before running, as well as access any _additional files_, enabling generators to
+introspect both user C# code and generator specific files.
+
+> **Note**: This proposal is separate from the [previous generator design](generators.md)
+
+### High Level Design Goals
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additional_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
+- Run _un-ordered_, each generator will see the same input compilation.
```

Comment:
💡 Consider explicitly stating that source generators are assumed to be deterministic with respect to their inputs. In other words, if the inputs to the compilation did not change and the output of the source generator was previously cached by the build, the source generator is not required to be run again for a build.